### PR TITLE
Update to v 0.3.3 -- more/better interpolation

### PR DESCRIPTION
**Pull request by ChrisBarker-NOAA**
Lots of commits, but it is mostly:

New / better interpolation
Updated to use new cellTree2D API.

and some demos of how to use that.

We have been using this version in production for a while, but would like to make it "official" so we can get a conda-forge build of it going.

### CODE REVIEW

**Comment by ayan-usgs** [comment]
I'll do my best to complete a review of all the changes this week.

My initial concern is that the travis tests are failing because of an import error related to the scipy package and that the scipy dependency is not represented in the requirements file.

**Comment by ayan-usgs** [comment]
Haven't forgotten about this. Just been super busy. :sweat_smile:

**Comment by ChrisBarker-NOAA** [comment]
me too :-)

I'll try to go in and clean up the dependencies, etc if I get a chance -- but not sure when.

**Comment by ChrisBarker-NOAA** [comment]
Whoo hoo! got all the tests to pass!!!
@ayan-usgs -- it should be ready to go.

@jahennen: congrats!

File: demos/inlet_heatmap.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+
+from netCDF4 import Dataset
+import numpy as np
+import pysgrid
+
+import matplotlib.pyplot as plt
+
+import cartopy.crs as ccrs
+# from cartopy.io import shapereader
```
Comment (ayan-usgs) [inline demos/inlet_heatmap.py]: Can this be removed?

File: pysgrid/sgrid.py [hunk 2]
```diff
@@ -419,29 +448,102 @@ def locate_faces(self,
...
-                               self._ind_memo_dict, _copy, _hash)
+                self._add_memo(points, res, grid, self._ind_memo_dict, _copy, _hash)
             return res
 
+    def locate_nearest(self,
+                       points,
+                       grid,
+                       _memo=False,
+                       _copy=False,
+                       _hash=None):
```
Comment (jay-hennen) [inline pysgrid/sgrid.py]: They are there to present a consistent API for locate_ functions, and are likely to be used in future.

File: demos/basic_interp.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ptsx, ptsy = np.mgrid[0:6:600j, 0:6:600j]
+pts = np.stack((ptsx, ptsy), axis=-1)
+
+
+interp_c = sgrid.interpolate_var_to_points(pts, c_var).reshape(600, 600)
+interp_e1 = sgrid.interpolate_var_to_points(pts, e1_var).reshape(600, 600).T
+interp_e2 = sgrid.interpolate_var_to_points(pts, e2_var).reshape(600, 600).T
+interp_n = sgrid.interpolate_var_to_points(pts, n_var).reshape(600, 600)
+
+import matplotlib.pyplot as plt
```
Comment (jay-hennen) [inline demos/basic_interp.py]: moved

File: demos/inlet_heatmap.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ang_ind = ind + [1, 1]
+angles = sgrid.angles[:][ang_ind[:, 0], ang_ind[:, 1]]
+u_rot, v_rot = rotate_vectors(interp_u, interp_v, angles)
+u_rot = u_rot.reshape(600, -1)
+v_rot = v_rot.reshape(600, -1)
+
+uv_vector_sum = vector_sum(u_rot, v_rot)
+
+
+import numpy as np
```
Comment (jay-hennen) [inline demos/inlet_heatmap.py]: moved

File: demos/inlet_heatmap.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+kw = dict(scale=1.0 / scale, pivot='middle', width=0.003, color='black')
+# q = plt.quiver(lon_data[::vscale, ::vscale], lat_data[::vscale, ::vscale],
+# u_rot[::vscale, ::vscale], v_rot[::vscale, ::vscale], zorder=2, **kw)
+
+cs = plt.pcolormesh(lon_data[::mscale, ::mscale],
+                    lat_data[::mscale, ::mscale],
+                    uv_vector_sum[::mscale, ::mscale], zorder=1, cmap=plt.cm.rainbow)
+ax.coastlines('10m')
+# plt.subplots(figsize=(20,20))
+# cs = plt.imshow(uv_vector_sum, cmap=plt.cm.rainbow)
```
Comment (jay-hennen) [inline demos/inlet_heatmap.py]: removed

File: pysgrid/sgrid.py [hunk 6]
```diff
@@ -347,8 +377,8 @@ def get_lines(self, grid='node'):
         """
         TEMPORARY
```
Comment (jay-hennen) [inline pysgrid/sgrid.py]: function removed

File: pysgrid/sgrid.py [hunk 7]
```diff
@@ -419,29 +448,102 @@ def locate_faces(self,
...
+        raw = np.ravel_multi_index(index.T, var.shape, mode='clip')
+        rv[:, 0] = np.take(var, raw)
+        raw += np.array(var.shape[1], dtype=np.int32)
+        rv[:, 1] = np.take(var, raw)
+        raw += 1
+        rv[:, 2] = np.take(var, raw)
+        raw -= np.array(var.shape[1], dtype=np.int32)
+        rv[:, 3] = np.take(var, raw)
+        return rv
+#         return np.ma.column_stack((var[x, y], var[x + 1, y], var[x + 1, y + 1], var[x, y + 1]))
```
Comment (jay-hennen) [inline pysgrid/sgrid.py]: agreed, removed

File: pysgrid/sgrid.py [hunk 8]
```diff
@@ -419,29 +448,102 @@ def locate_faces(self,
...
+            self._kd_trees = {'node': None,
+                              'edge1': None,
+                              'edge2': None,
+                              'center': None}
+        lon, lat = self._get_grid_vars(grid)
+        if lon is None or lat is None:
+            raise ValueError(
+                "{0}_lon and {0}_lat must be defined in order to create and use KDTree for this grid".format(grid))
+        lin_points = np.column_stack((lon.ravel(), lat.ravel()))
+        print(lin_points.shape)
```
Comment (jay-hennen) [inline pysgrid/sgrid.py]: no, removed

File: pysgrid/sgrid.py [hunk 9]
```diff
@@ -451,31 +553,64 @@ def build_celltree(self, grid='node'):
...
+            variable = variable[slices]
+        if len(variable.shape) > 2:
+            raise ValueError("Variable has too many dimensions to \
+            associate with grid. Please specify slices.")
 
+        ind = indices.copy() - [yslice.start, xslice.start]
+        result = self.get_variable_at_index(variable, ind)
+        return result
+
+#     @profile
```
Comment (jay-hennen) [inline pysgrid/sgrid.py]: removed

File: pysgrid/sgrid.py [hunk 10]
```diff
@@ -659,86 +729,82 @@ def x_to_l(x, y, a, b):
...
                 m2 = (-bb + det) / (2 * aa)
-                l2 = (x[ind_arr] - a[0][ind_arr] - a[2][ind_arr] *
-                      m2) / (a[1][ind_arr] + a[3][ind_arr] * m2)
+                l2 = (x[t] - a[0][t] - a[2][t] * 
+                      m2) / (a[1][t] + a[3][t] * m2)
 
-                m[ind_arr] = m1
-                l[ind_arr] = l1
+#                 m[t] = m1
+#                 l[t] = l1
```
Comment (jay-hennen) [inline pysgrid/sgrid.py]: removed

File: pysgrid/tests/test_interpolation.py [hunk 11]
```diff
@@ -6,8 +6,10 @@
 
 import numpy as np
 
-from ..sgrid import SGrid
-from ..utils import points_in_polys
+# from ..sgrid import SGrid
+# from ..utils import points_in_polys
```
Comment (jay-hennen) [inline pysgrid/tests/test_interpolation.py]: removed

File: pysgrid/tests/test_interpolation.py [hunk 12]
```diff
@@ -118,3 +120,22 @@ def test_points_in_polys2():
     answer = sgrid.locate_faces(boundaries + 1, 'node') == [0, 0]
     answer = np.logical_and(answer[:, 0], answer[:, 1])
     assert (answer == pinp).all()
+
+def test_nearest_neighbor():
+    sgrid.build_kdtree()
+
+    diagonal = [[0, 0], [1, 1], [1, 2], [1, 2.1], [2.1,2.1], [5, 5], [6, 6]]
+    inds = sgrid.locate_nearest(diagonal, 'node')
+    print inds
```
Comment (jay-hennen) [inline pysgrid/tests/test_interpolation.py]: removed

File: pysgrid/tests/test_interpolation.py [hunk 13]
```diff
@@ -118,3 +120,22 @@ def test_points_in_polys2():
...
+                        [0, 0],
+                        [0, 0],
+                        [1, 0],
+                        [1, 1],
+                        [2, 2],
+                        [2, 2]], dtype=np.int64)
+
+    assert np.all(inds ==ind_ans)
+
+if __name__ == '__main__':
```
Comment (jay-hennen) [inline pysgrid/tests/test_interpolation.py]: No, but it is also harmless. Removed anyway

File: pysgrid/sgrid.py [hunk 14]
```diff
@@ -599,51 +713,7 @@ def interpolation_alphas(self,
                              grid='center',
                              _memo=False,
                              _copy=False,
-                             _hash=None,):
-        """
-        Given an array of points, this function will return the bilinear
```
Comment (jay-hennen) [inline pysgrid/sgrid.py]: Agreed, though it's going to take longer than I have time for now. In the interim I've added a top level summary of what the function does and the sources the math is based on, and some notes in the code for future reference

File: pysgrid/tests/test_interpolation.py (hunk 12 above)
Comment (jay-hennen) [inline pysgrid/tests/test_interpolation.py]: removed

**Comment by ayan-usgs** [comment]
@jay-hennen: Thanks for making those changes.

Tests are failing because of: `./pysgrid/tests/test_interpolation.py:136:35: W292 no newline at end of file`. That's pretty superficial, so I'm going to proceed with the merge.

**Pull request merged**